### ISSUE TITLE
Keep BottomNavigationBar When Push to New Screen with Navigator

### ISSUE THREAD

**Issue by anotherHarry** [flutter/flutter#16181]
Hi, I don't know to ask for help for implementation so I fire an issue. If you find this not appropriate, please delete. Thanks.

In iOS, we have a `UITabBarController` which stays permanently at the bottom of the screen when we push to a new ViewController.

In iOS, we have a UITabBarController which stays permanently at the bottom of the screen when we push to a new ViewController.

In Flutter, we have a `bottomNavigationBar` of a `Scaffold`. However, unlike iOS, when we `Navigator.push` to a new screen, this `bottomNavigationBar` disappears.

In my app, I want to fulfil this requirement: Home screen has a `bottomNavigationBar` with 2 items (*a* & *b*) presenting screen *A* & *B*. By default, screen *A* is displayed. Inside screen *A*, there is a button. Tap that button,  `Navigator.push` to screen *C*. Now in screen *C*, we can still see the `bottomNavigationBar`. Tap item *b*, I go to screen *B*. Now in screen *B*, tap item *a* in the `bottomNavigationBar`, I go back to screen *C* (not *A*, *A* is currently below *C* in the navigation hierarchy). 

How can I do this? Thanks, guys.

I'm including some pictures for demonstration:

Screen A
https://i.stack.imgur.com/TLckw.png

Tap Go to C button, push to screen C
https://i.stack.imgur.com/5MpU5.png

Tap Right item inside bottom navigation bar, go to screen B
https://i.stack.imgur.com/gtqQq.png

**Comment by xster** [comment]
There's no premade widget that does exactly what you're describing since it's not a standard Material Design pattern. There is a [CupertinoTabScaffold](https://docs.flutter.io/flutter/cupertino/CupertinoTabScaffold-class.html) which has the desired pattern if you're ok with iOS aesthetics. Otherwise, it shouldn't be difficult to look at the CupertinoTabScaffold code to craft a custom widget that reuses the Material BottomNavigationBar

**Comment by anotherHarry** [comment]
@xster Your answer is correct. I just take a look at CupertinoTabScaffold and it seems to have potential. I will try implement it and update this thread further.

**Comment by anotherHarry** [comment]
tl;dr: Use CupertinoTabBar with CupertinoTabScaffold

The problem is not in Flutter but in UX.

It turned out Material Design doesn't recommend sub-pages in the hierarchy to access the Bottom navigation bar.

However, iOS Human Interface Guide recommend this. So, to use this feature, I had to adapt Cupertino widgets instead of Material ones. Specifically, in main, return a WidgetsApp/MaterialApp which contains a CupertinoTabScaffold. Implement the tab bar with a  CupertinoTabBar and each screen is a CupertinoTabView.

**Comment by brandondiamond** [comment]
I'm generally interested in "pushing" and "popping" regions (i.e., widgets) in the UI. iOS does something like this to implement things like tabbar and other parent view controllers.

How can we achieve the same effect in Flutter? Is there a concise pattern or an example that might be helpful to a beginner?

Examples would be "pushing" a modal over a portion of the screen, or keeping a tab bar around between pushes and pops.

**Comment by amrnt** [comment]
@xster Regarding this, I can see that the current BottomNavigationBar UX implementation doesn't really implement the standards: https://material.io/design/components/bottom-navigation.html#behavior

I can see that the implementation of CupertinoTab implements some of the behaviours better. For example, each "Tab" has its own routing state.

I opened an issue sometime ago to mimic one of the behaviours (It takes the user to the screen associated with it): https://github.com/flutter/flutter/issues/15493

**Comment by xster** [comment]
@amrnt: https://material.io/design/components/bottom-navigation.html#behavior is a new behavior from Material 2. Please file a feature request for the new behaviour in the material BottomNavigationBar

cc @HansMuller, I think we chatted about this a while back.

**Comment by amrnt** [comment]
@xter @HansMuller The bad UX in Flutter's documentation, that it has link to [the old link] for Bottom Navigation, that redirects to the new link @ material.io.

Users would expect that the implementation reflects on the material.io description  of the component. 

I'm not sure as well if there is other components that breaks after Material v2 released.

I will open a feature request for BottomNavigation.

**Comment by ericjatis** [comment]
So best practice on Material is to have a repeated BottomNavigationBar on those top 5 screen? In case I don't want CupertinoTabBar

**Comment by amrnt** [comment]
@neckaros here's the open issue: https://github.com/flutter/flutter/issues/18740 (i see no progress)

Or build it yourself read comments from here: https://github.com/flutter/flutter/issues/15493#issuecomment-410299617

**Comment by neckaros** [comment]
Thanks @amrnt 
I did implement it myself indeed adapting the code from Cupertino widgets 😁

**Comment by ermissa** [comment]
Hi, I am facing with same problem. Can there be some news about this issue that I have missed?

The well known apps (e.g youtube, spotify, reddit) which use bottom navigation bars are all working with the same logic. I think there should be a standart way to do this in Flutter.

**Comment by premtemp1** [comment]
I have the same issue is the only option to repeat the footer on each page?

**Comment by DaniyarGilimov** [comment]
this has good solutions
https://github.com/flutter/flutter/issues/31483

**Comment by amrnt** [comment]
@DaniyarGilimov IndexedStack or Stack is a solution but not the ultimate solution.

The base code should follow the specs of material design from the the square zero. New comers when they start using flutter, they expect that things would reflect what is on https://material.io. Not the case for many things and the `BottomNavigationBar` is one of them. See here: https://github.com/flutter/flutter/issues/18740

**Comment by DaniyarGilimov** [comment]
@amrnt totally agree with you

**Comment by jingjiewei** [comment]
I think the easiest way is to duplicate the bottom navigation(make it a stateless widget) in each screen and overwrite the default page navigating animation(so it looks stick, no jumping)

**Comment by HansMuller** [comment]
@jingjiewei @DaniyarGilimov - You might want to check out https://github.com/flutter/flutter/issues/18740#issuecomment-508134439

Also: although there's nothing wrong with commenting on closed issues, if you're trying to get the Flutter team's attention, then it's best to comment on an open issue or open a new one.

**Comment by ayham95** [comment]
Here’s a great efficient solution!
https://pub.dev/packages/custom_navigator

**Comment by Ec7i9se** [comment]
You need just wrap your widget after Scaffold into new Navigator. Check this answer [stackoverflow](https://stackoverflow.com/a/65666020/9852658) make sure you use proper context when pushing
